syntax = "proto2";

import "serialization/geometry.proto";
import "serialization/integrators.proto";
import "serialization/physics.proto";

package principia.serialization;

option cc_enable_arenas = true;

message GravityModel {
  message Body {
    required string name = 1;
    // Massive body.
    oneof massive {
      string gravitational_parameter = 2;
      string mass = 13;
    }
    // Rotating body.
    optional string reference_instant = 12;  // Terrestrial time.
    optional string min_radius = 15;
    optional string mean_radius = 7;
    optional string max_radius = 16;
    optional string axis_right_ascension = 3;
    optional string axis_declination = 4;
    optional string reference_angle = 8;
    optional string angular_frequency = 9;
    // Oblate body.
    optional string reference_radius = 6;
    oneof oblateness {
      double j2 = 11;  // Unnormalized.
      OblateBody.Geopotential geopotential = 14;
    }
    reserved 5, 10;
  }
  oneof frame {
    Frame.SolarSystemTag solar_system_frame = 1;
    Frame.PluginTag plugin_frame = 3;
  }
  repeated Body body = 2;
}

message InitialState {
  message Cartesian {
    message Body {
      required string name = 1;
      required string x = 2;
      required string y = 3;
      required string z = 4;
      required string vx = 5;
      required string vy = 6;
      required string vz = 7;
    }
    repeated Body body = 1;
  }
  message Keplerian {
    message Body {
      message Elements {
        required double eccentricity = 1;
        oneof category2 {  // See kepler_orbit.hpp.
          string semimajor_axis = 2;
          string mean_motion = 3;
          string period = 8;
        }
        required string inclination = 4;
        required string longitude_of_ascending_node = 5;
        required string argument_of_periapsis = 6;
        required string mean_anomaly = 7;
      }
      required string name = 1;
      optional string parent = 2;
      optional Elements elements = 3;
    }
    repeated Body body = 1;
  }
  oneof frame {
    Frame.SolarSystemTag solar_system_frame = 1;
    Frame.PluginTag plugin_frame = 6;
  }
  required string epoch = 7;  // Terrestrial time.
  oneof coordinates {
    Cartesian cartesian = 4;
    Keplerian keplerian = 5;
  }
  reserved 2, 3;
  reserved "body";
}

message NumericsBlueprint {
  message Ephemeris {
    required FixedStepSizeIntegrator.Kind integrator = 1;
    required string step = 2;
    required string fitting_tolerance = 3;
    required string geopotential_tolerance = 4;
  }
  message History {
    required FixedStepSizeIntegrator.Kind integrator = 1;
    required string step = 2;
  }
  message Psychohistory {
    required AdaptiveStepSizeIntegrator.Kind integrator = 1;
    required string length_integration_tolerance = 2;
    required string speed_integration_tolerance = 3;
  }
  required Ephemeris ephemeris = 1;
  required History history = 2;
  required Psychohistory psychohistory = 3;
}

message SolarSystemFile {
  oneof file {
    GravityModel gravity_model = 1;
    InitialState initial_state = 2;
    NumericsBlueprint numerics_blueprint = 3;
  }
}
